### PR TITLE
Make JsonEventLogger compatible with HDFS-like FileSystems

### PR DESCRIPTION
**Pull request by yaooqinn**
### _Why are the changes needed?_

FSDataOutputStream is used for this
### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @pan3793 @ulysses-you @zhang1002

File: externals/kyuubi-spark-sql-engine/pom.xml [hunk 1]
```diff
@@ -68,6 +68,19 @@
...
+        <dependency>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-client-api</artifactId>
+            <scope>provided</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-client-runtime</artifactId>
+            <scope>provided</scope>
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/pom.xml:81]: keep `hadoop-client-runtime` in test scope

File: externals/kyuubi-spark-sql-engine/pom.xml [hunk 2]
```diff
@@ -68,6 +68,19 @@
             <scope>provided</scope>
         </dependency>
 
+
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/pom.xml:71]: nit: keep one blank line

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/JsonEventLogger.scala [hunk 3]
```diff
@@ -33,32 +36,51 @@ import org.apache.kyuubi.service.AbstractService
  * The {eventType} is based on core concepts of the Kyuubi systems, e.g. engine/session/statement
  * @param logName the engine id formed of appId + attemptId(if any)
  */
-class JsonEventLogger(logName: String)
+class JsonEventLogger(logName: String, hadoopConf: Configuration)
   extends AbstractService("JsonEventLogger") with EventLogger with Logging {
 
-  private var logRoot: Path = _
-  private val writers = new scala.collection.mutable.HashMap[String, PrintWriter]()
+  type Logger = (PrintWriter, Option[FSDataOutputStream])
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/JsonEventLogger.scala:42]: How about `type Logger = Either[PrintWriter, FSDataOutputStream]`

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/JsonEventLogger.scala [hunk 4]
```diff
@@ -33,32 +36,51 @@ import org.apache.kyuubi.service.AbstractService
...
+
+  private var logRoot: URI = _
+  private var fs: FileSystem = _
+  private val writers =
+    new scala.collection.mutable.HashMap[String, Logger]()
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/JsonEventLogger.scala:47]: `scala.collection.mutable.HashMap` => `mutable.HashMap`

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/JsonEventLogger.scala (hunk 3 above)
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/JsonEventLogger.scala:42]: PrintWriter is not optional

**Review by pan3793 — APPROVED** [review]

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/JsonEventLogger.scala [hunk 5]
```diff
@@ -33,32 +39,50 @@ import org.apache.kyuubi.service.AbstractService
...
+      val eventPath = new Path(new Path(logRoot), event.eventType)
+      FileSystem.mkdirs(fs, eventPath, JSON_LOG_DIR_PERM)
+      val logFile = new Path(eventPath, logName + ".json")
+      var hadoopDataStream: FSDataOutputStream = null
+      val rawStream = if (logFile.toUri.getScheme == "file") {
+        new FileOutputStream(logFile.toUri.getPath)
+      } else {
+        hadoopDataStream = fs.create(logFile)
+        hadoopDataStream
+      }
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/JsonEventLogger.scala:62]: why we need distinguish the scheme here, we can't just do this ?
```
val stream = fs.create(logFile)
val bStream = new BufferedOutputStream(stream)
```

Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/JsonEventLogger.scala:62]: avoid hdfs bug

Comment (ulysses-you) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/events/JsonEventLogger.scala:62]: Can we add that bug link here

**Pull request closed without merging**